storage/objects/stream_file_upload.go

```diff
@@ -43,8 +43,21 @@ func streamFileUpload(w io.Writer, bucket, object string) error {
 	ctx, cancel := context.WithTimeout(ctx, time.Second*50)
 	defer cancel()
 
+	o := client.Bucket(bucket).Object(object)
+
+	// Add preconditions so that the upload may safely retry on transient errors.
```

Comment:
IMO the main reason to add a precondition is to avoid race conditions and data corruption as described here https://cloud.google.com/storage/docs/request-preconditions#examples_of_race_conditions_and_data_corruption . Safe retries are a side effect. Maybe a subtle distinction but I think we should encourage using a precondition on its own terms rather than emphasizing the connection with retries here.